tests/unit/s2n_self_talk_session_resumption_test.c

```diff
@@ -0,0 +1,286 @@
+    EXPECT_SUCCESS(s2n_session_ticket_get_data(ticket, cb_session_data_len, cb_session_data));
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_setup_test_ticket_key(struct s2n_config *config)
+{
+    POSIX_ENSURE_REF(config);
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc5869#appendix-A.1
+     *# PRK  = 0x077709362c2e32df0ddc3f0dc47bba63
+     *#        90b6c73bb50f9c3122ec844ad7c2b3e5 (32 octets)
+     **/
+    S2N_BLOB_FROM_HEX(ticket_key,
+    "077709362c2e32df0ddc3f0dc47bba63"
+    "90b6c73bb50f9c3122ec844ad7c2b3e5");
+
+    /* Set up encryption key */
+    uint64_t current_time;
+    uint8_t ticket_key_name[16] = "2016.07.26.15\0";
+    EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(config, 1));
+    EXPECT_SUCCESS(config->wall_clock(config->sys_clock_ctx, &current_time));
+    EXPECT_SUCCESS(s2n_config_add_ticket_crypto_key(config, ticket_key_name, strlen((char *)ticket_key_name),
+                    ticket_key.data, ticket_key.size, current_time/ONE_SEC_IN_NANOS));
+
+    return S2N_SUCCESS;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Setup server config */
+    struct s2n_config *server_config = s2n_config_new();
+    EXPECT_NOT_NULL(server_config);
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(server_config, "default_tls13"));
+    EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(server_config));
+    struct s2n_cert_chain_and_key *chain_and_key = NULL;
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key, S2N_ECDSA_P384_PKCS1_CERT_CHAIN,
+                                                   S2N_ECDSA_P384_PKCS1_KEY));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, chain_and_key));
+    struct s2n_cert_chain_and_key *tls12_chain_and_key = NULL;;
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&tls12_chain_and_key, S2N_DEFAULT_TEST_CERT_CHAIN, 
+                                                   S2N_DEFAULT_TEST_PRIVATE_KEY));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, tls12_chain_and_key));
+    EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(server_config, true));
+    EXPECT_SUCCESS(s2n_setup_test_ticket_key(server_config));
+
+    /* Setup TLS1.3 client config */
+    struct s2n_config *tls13_client_config = s2n_config_new();
+    EXPECT_NOT_NULL(tls13_client_config);
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(tls13_client_config, "default_tls13"));
+    EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(tls13_client_config));
+    EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(tls13_client_config, true));
+    EXPECT_SUCCESS(s2n_config_set_session_ticket_cb(tls13_client_config, s2n_test_session_ticket_cb, NULL));
+
+    /* Setup TLS1.2 client config */
+    struct s2n_config *tls12_client_config = s2n_config_new();
+    EXPECT_NOT_NULL(tls12_client_config);
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(tls12_client_config, "20170210"));
```

Comment:
We could use cipher preference `default` here instead as it is the same as `20170210` but easier to read.